common/src/main/java/com/scottlogic/deg/generator/fieldspecs/whitelist/FrequencyDistributedSet.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2019 Scott Logic Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.scottlogic.deg.generator.fieldspecs.whitelist;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class FrequencyDistributedSet<T> implements DistributedSet<T> {
+
+    private static final FrequencyDistributedSet<?> EMPTY = new FrequencyDistributedSet<>(Collections.emptySet());
+
+    private final Set<WeightedElement<T>> underlyingWeights;
+
+    private final List<WeightedElement<T>> underlyingCumulativeWeights;
+
+    public FrequencyDistributedSet(final Set<WeightedElement<T>> underlyingWeights) {
+        if (underlyingWeights.isEmpty()) {
+            this.underlyingWeights = Collections.emptySet();
+            this.underlyingCumulativeWeights = generateCumulative(underlyingWeights);
+        } else {
+            if (underlyingWeights.contains(null)) {
+                throw new IllegalArgumentException("DistributedSet should not contain null elements");
+            }
+
+            double total = total(underlyingWeights);
+            Set<WeightedElement<T>> normalisedWeights = underlyingWeights.stream()
+                .map(holder -> new WeightedElement<>(holder.element(), holder.weight() / total))
+                .collect(Collectors.toSet());
+            this.underlyingWeights = normalisedWeights;
+            this.underlyingCumulativeWeights = generateCumulative(normalisedWeights);
+        }
+    }
+
+    public static <T> FrequencyDistributedSet<T> singleton(final T element) {
```

Comment:
not sure this is worth having, it's only called once in real code

Reply:
actually. it could also be called in fieldSpecHelper.

i don't see why this should use the uniform method though. that seems like an unneccessary mapping

Reply:
`singleton` is a special case of `uniform`.

Reply:
yes, but it doesn't need to be

Reply:
What do you propose we put here instead?